### ISSUE TITLE
target with 'no-remote-exec' is not cached remotely

### ISSUE THREAD

**Issue by ishikhman** [bazelbuild/bazel#9412]
The following test is failing at the last check:

```
function test_tag_no_remote_exec() {
  mkdir -p a
  cat > a/BUILD <<'EOF'
genrule(
  name = "foo",
  srcs = [],
  outs = ["foo.txt"],
  cmd = "echo \"foo\" > \"$@\"",
  tags = ["no-remote-exec"]
)
EOF

  bazel build \
    --spawn_strategy=remote,local \
    --remote_executor=grpc://localhost:${worker_port} \
    //a:foo >& $TEST_log || "Failed to build //a:foo"

  expect_log "1 local"
  expect_not_log "1 remote"

  bazel clean

  bazel build \
    --spawn_strategy=remote,local \
    --remote_executor=grpc://localhost:${worker_port} \
    //a:foo >& $TEST_log || "Failed to build //a:foo"

  expect_not_log "1 remote"
  expect_log "remote cache hit"
}
```

**When this is the case?**
When from the very beginning bazel selects non-local strategy for a target:
- if `tags = ["no-remote-exec"]`
- if  `tags = ["no-remote"]`

**Comment by Qinusty** [comment]
Which version of bazel does this affect and is it fixed in the 3.0 release candidates?

**Comment by Qinusty** [comment]
After looking into this for a little bit, this seems to be caused by the combination of the effects of the `--remote_executor=` flag and the tag `no-remote-exec`.

From what I can see, the `--remote-executor=` flag forces the `BlazeExecutor` class to be instantiated with it's spawn cache configured as `NoSpawnCache`. 

Additionally, the use of `no-remote-exec` as a tag forces the `ProxySpawnActionContext.resolve()` function to filter out `RemoteSpawnStrategy` as a possible spawn execution strategy for the action. 

So this leaves us with a Standalone/Sandboxed runner etc, making use of the `AbstractSpawnStrategy` logic for caching (The only alternative to this is the `RemoteSpawnRunner` class which handles remote caching in relation to remote exection).

The `AbstractSpawnStrategy` caching logic depends on the `SpawnCache` for the executor being configured as part of it's ContextMap. All in all, this leaves `no-remote-exec` actions falling into a code path without any remote caching configuration.

I'll continue looking into this, just thought I'd document my findings so far.

**NOTE**: All of this analysis was done on bazel version 1.2.1 but I have confirmed this issue exists in version 2.0 (Unsure about 3.0).

**Comment by Qinusty** [comment]
This should be fixed by https://github.com/bazelbuild/bazel/pull/10964 which is available in Bazel 3.2.0rc1.

**Comment by coeuvre** [comment]
Closing. Feel free to re-open if this is still an issue.
